framework/src/utils/PerfGraph.C

```diff
@@ -0,0 +1,447 @@
+    mooseError("Unknown section_name: ", section_name, " in PerfGraph::getTime() ");
+
+  auto app_time = _section_time_ptrs[0]->_total;
+
+  switch (type)
+  {
+    case SELF:
+      return section_it->second._self;
+    case CHILDREN:
+      return section_it->second._children;
+    case TOTAL:
+      return section_it->second._total;
+    case SELF_AVG:
+      return section_it->second._self / static_cast<Real>(section_it->second._num_calls);
+    case CHILDREN_AVG:
+      return section_it->second._children / static_cast<Real>(section_it->second._num_calls);
+    case TOTAL_AVG:
+      return section_it->second._total / static_cast<Real>(section_it->second._num_calls);
+    case SELF_PERCENT:
+      return 100. * (section_it->second._self / app_time);
+    case CHILDREN_PERCENT:
+      return 100. * (section_it->second._children / app_time);
+    case TOTAL_PERCENT:
+      return 100. * (section_it->second._total / app_time);
+    default:
+      ::mooseError("Unknown TimeType");
+  }
+}
+
+void
+PerfGraph::push(const PerfID id)
+{
+  if (!_active)
+    return;
+
+  auto new_node = _stack[_current_position]->getChild(id);
+
+  // Set the start time
+  new_node->setStartTime(std::chrono::steady_clock::now());
+
+  // Increment the number of calls
+  new_node->incrementNumCalls();
+
+  _current_position++;
+
+  if (_current_position >= MAX_STACK_SIZE)
+    mooseError("PerfGraph is out of stack space!");
+
+  _stack[_current_position] = new_node;
+}
+
+void
+PerfGraph::pop()
+{
+  if (!_active)
+    return;
+
+  _stack[_current_position]->addTime(std::chrono::steady_clock::now());
+
+  _current_position--;
+}
+
+void
+PerfGraph::updateTiming()
+{
+  // First update all of the currently running nodes
+  auto now = std::chrono::steady_clock::now();
+  for (unsigned int i = 0; i <= _current_position; i++)
+  {
+    auto node = _stack[i];
+    node->addTime(now);
+    node->setStartTime(now);
+  }
+
+  // Zero out the entries
+  for (auto & section_time_it : _section_time)
+  {
+    auto & section_time = section_time_it.second;
+
+    section_time._num_calls = 0;
+    section_time._self = 0.;
+    section_time._children = 0.;
+    section_time._total = 0.;
+  }
+
+  recursivelyFillTime(_root_node.get());
+
+  // Update vector pointing to section times
+  // Note: we are doing this _after_ recursively filling
+  // because new entries may have been created
+  _section_time_ptrs.resize(_id_to_section_name.size());
+
+  for (auto & section_time_it : _section_time)
+  {
+    auto id = _section_name_to_id[section_time_it.first];
+
+    _section_time_ptrs[id] = &section_time_it.second;
+  }
+}
+
+void
+PerfGraph::recursivelyFillTime(PerfNode * current_node)
+{
+  auto id = current_node->id();
+
+  auto self = std::chrono::duration<double>(current_node->selfTime()).count();
+  auto children = std::chrono::duration<double>(current_node->childrenTime()).count();
+  auto total = std::chrono::duration<double>(current_node->totalTime()).count();
```

Comment:
Just confirming that you want `double`. I believe this is correct.

Reply:
Yeah - `double` is how you get seconds out.  I don't think we actually want this to change with `Real`